### PR TITLE
pixel count for the pixel classification in QA

### PR DESCRIPTION
**Pull request by seongsujeong**
The code still needs a littlee bit of work, but please feel free to post comments.

### CODE REVIEW

File: src/compass/s1_cslc_qa.py [hunk 1]
```diff
@@ -466,3 +478,97 @@ def write_qa_dicts_to_json(self, file_path):
...
+            Returns
+            -------
+            ratio_valid_pixel_land:  float
+                Ratio of valid pixels on land
+        '''
+
+        drv_shp_in = ogr.GetDriverByName('GPKG')
+        if os.path.exists(LAND_GPKG_FILE):
+            coastline_shapefile = drv_shp_in.Open(LAND_GPKG_FILE)
+        else:
```
Comment (vbrancat) [inline src/compass/s1_cslc_qa.py:504]: I think here it would be more appropriate to have a `try`/`except` instead of a `if`/`else`

File: src/compass/s1_cslc_qa.py [hunk 2]
```diff
@@ -466,3 +481,133 @@ def write_qa_dicts_to_json(self, file_path):
...
+            shapefile_path: str
+                Path to the coastline shapefile
+
+
+            Returns
+            -------
+            ratio_valid_pixel_land:  float
+                Ratio of valid pixels on land
+        '''
+
```
Comment (LiangJYu) [inline src/compass/s1_cslc_qa.py:503]:
```suggestion
```
formatting nit

File: src/compass/s1_cslc_qa.py (hunk 1 above)
Comment (LiangJYu) [reply src/compass/s1_cslc_qa.py:507]: I think this ok. The rest of the code uses `if/else` for path checking.

File: src/compass/s1_cslc_qa.py [hunk 3]
```diff
@@ -466,3 +481,133 @@ def write_qa_dicts_to_json(self, file_path):
...
+        mask_valid_inside_burst = mask_geocoded_burst & ~np.isnan(cslc_array)
+
+        mask_valid_land_pixel = mask_geocoded_burst & mask_land
+
+        percent_valid_land_px = mask_valid_land_pixel.sum() / mask_geocoded_burst.sum() * 100
+        percent_valid_px = mask_valid_inside_burst.sum() / mask_geocoded_burst.sum() * 100
+
+        return percent_valid_land_px, percent_valid_px
+
+
```
Comment (LiangJYu) [inline src/compass/s1_cslc_qa.py:550]:
```suggestion
```
formatting nit

File: src/compass/s1_cslc_qa.py [hunk 4]
```diff
@@ -466,3 +481,133 @@ def write_qa_dicts_to_json(self, file_path):
...
+        percent_combined = mask_combined_inside_burst.sum() / mask_geocoded_burst.sum() * 100
+
+        return percent_shadow, percent_layover, percent_combined
+
+
+def _get_valid_pixel_mask(arr_cslc):
+    # temp. code to test for edge cases
+    #arr_cslc[2000:2500, 9000:10000] = (1 + 1j)*np.nan
+    #arr_cslc[2500:, 0:7500] = (1 + 1j)
+
```
Comment (LiangJYu) [inline src/compass/s1_cslc_qa.py:591]:
```suggestion
```
formatting nit

File: src/compass/s1_cslc_qa.py [hunk 5]
```diff
@@ -466,3 +481,133 @@ def write_qa_dicts_to_json(self, file_path):
...
+            shapefile_path: str
+                Path to the coastline shapefile
+
+
+            Returns
+            -------
+            ratio_valid_pixel_land:  float
+                Ratio of valid pixels on land
+        '''
+
```
Comment (LiangJYu) [inline src/compass/s1_cslc_qa.py:570]:
```suggestion
```
formatting nit

File: src/compass/s1_cslc_qa.py [hunk 6]
```diff
@@ -466,3 +481,133 @@ def write_qa_dicts_to_json(self, file_path):
...
+        layover_shadow_mask_array = np.array(cslc_h5py_root[f'{DATA_PATH}/layover_shadow_mask'])
```
Comment (LiangJYu) [inline src/compass/s1_cslc_qa.py:571]:
```suggestion
        layover_shadow_mask_array = cslc_h5py_root[f'{DATA_PATH}/layover_shadow_mask'][()]
```
`[()]` should return `np.ndarray`?

File: src/compass/s1_cslc_qa.py [hunk 7]
```diff
@@ -466,3 +481,133 @@ def write_qa_dicts_to_json(self, file_path):
...
+        mask_geocoded_burst = layover_shadow_mask_array != 127
+
+        mask_shadow_inside_burst = mask_geocoded_burst & (layover_shadow_mask_array == 1)
+        percent_shadow = mask_shadow_inside_burst.sum() / mask_geocoded_burst.sum() * 100
+
+        mask_layover_inside_burst = mask_geocoded_burst & (layover_shadow_mask_array == 2)
+        percent_layover = mask_layover_inside_burst.sum() / mask_geocoded_burst.sum() * 100
+
+        mask_combined_inside_burst = mask_geocoded_burst & (layover_shadow_mask_array == 3)
+        percent_combined = mask_combined_inside_burst.sum() / mask_geocoded_burst.sum() * 100
```
Comment (LiangJYu) [inline src/compass/s1_cslc_qa.py:582]:
```suggestion
        n_unmasked_pxls = mask_geocoded_burst.sum()
        
        mask_shadow_inside_burst = mask_geocoded_burst & (layover_shadow_mask_array == 1)
        percent_shadow = mask_shadow_inside_burst.sum() / n_unmasked_pxls * 100

        mask_layover_inside_burst = mask_geocoded_burst & (layover_shadow_mask_array == 2)
        percent_layover = mask_layover_inside_burst.sum() / n_unmasked_pxls * 100

        mask_combined_inside_burst = mask_geocoded_burst & (layover_shadow_mask_array == 3)
        percent_combined = mask_combined_inside_burst.sum() / n_unmasked_pxls * 100
```
Avoid repeat calls to `sum()` by caching sum of unmasked pixels.

File: src/compass/s1_cslc_qa.py [hunk 8]
```diff
@@ -466,3 +481,133 @@ def write_qa_dicts_to_json(self, file_path):
...
+    def compute_layover_shadow_pixel_percent(self, cslc_h5py_root):
+        '''
+            Compute the ratio of valid pixels on land area
+
+            Parameters
+            ----------
+            cslc_h5py_path: h5py.File
+                Root of the CSLC-S1 HDF5 product
+            shapefile_path: str
+                Path to the coastline shapefile
```
Comment (LiangJYu) [inline src/compass/s1_cslc_qa.py:562]:
```suggestion
            Compute the percentage of layover, shadow, and layover/shadow pixels in
            geocoded radar grid area

            Parameters
            ----------
            cslc_h5py_root: h5py.File
                Root of the CSLC-S1 HDF5 product
```
Description and parameters need to be updated

File: src/compass/s1_cslc_qa.py [hunk 9]
```diff
@@ -3,18 +3,27 @@
...
 from compass.s1_rdr2geo import (file_name_los_east,
                                 file_name_los_north,file_name_local_incidence,
                                 file_name_x, file_name_y, file_name_z)
 from compass.utils.h5_helpers import (DATA_PATH, METADATA_PATH, TIME_STR_FMT,
                                       QA_PATH, add_dataset_and_attrs, Meta)
 
 
+# determine the path to the world land GPKG file
+from compass import __path__ as compass_path
+LAND_GPKG_FILE = os.path.join(compass_path[0], 'data', 'GSHHS_l_L1.shp.no_attrs.gpkg')
```
Comment (LiangJYu) [inline src/compass/s1_cslc_qa.py:25]:
```
LAND_GPKG_FILE = os.path.join(WORKFLOW_SCRIPTS_DIR, 'data', 'GSHHS_l_L1.shp.no_attrs.gpkg')
```
If you add `from compass.utils.helpers import WORKFLOW_SCRIPTS_DIR`

File: src/compass/s1_cslc_qa.py [hunk 10]
```diff
@@ -466,3 +481,133 @@ def write_qa_dicts_to_json(self, file_path):
...
+        mask_combined_inside_burst = mask_geocoded_burst & (layover_shadow_mask_array == 3)
+        percent_combined = mask_combined_inside_burst.sum() / mask_geocoded_burst.sum() * 100
+
+        return percent_shadow, percent_layover, percent_combined
+
+
+def _get_valid_pixel_mask(arr_cslc):
+    # temp. code to test for edge cases
+    #arr_cslc[2000:2500, 9000:10000] = (1 + 1j)*np.nan
+    #arr_cslc[2500:, 0:7500] = (1 + 1j)
```
Comment (vbrancat) [inline src/compass/s1_cslc_qa.py:590]: @seongsujeong please remove this comment

File: src/compass/s1_cslc_qa.py [hunk 11]
```diff
@@ -243,12 +252,15 @@ def shadow_pixel_classification(self, cslc_h5py_root):
         cslc_h5py_root: h5py.File
             Root of CSLC HDF5
         '''
+
```
Comment (LiangJYu) [inline src/compass/s1_cslc_qa.py:255]:
```suggestion
```
formatting nit

File: src/compass/s1_cslc_qa.py [hunk 12]
```diff
@@ -466,3 +481,133 @@ def write_qa_dicts_to_json(self, file_path):
...
+    def compute_valid_land_pixel_percent(self, cslc_h5py_root, pol):
+        '''
+            Compute the ratio of valid pixels on land area
+
+            Parameters
+            ----------
+            cslc_h5py_path: h5py.File
+                Root of the CSLC-S1 HDF5 product
+            shapefile_path: str
+                Path to the coastline shapefile
```
Comment (LiangJYu) [inline src/compass/s1_cslc_qa.py:495]:
```suggestion
            pol: str
                Polarization of CSLC
```

File: src/compass/s1_cslc_qa.py [hunk 13]
```diff
@@ -466,3 +481,133 @@ def write_qa_dicts_to_json(self, file_path):
...
+        mask_land = rasterized_land.ReadAsArray()
+
+        mask_geocoded_burst = _get_valid_pixel_mask(cslc_array)
+
+        mask_valid_inside_burst = mask_geocoded_burst & ~np.isnan(cslc_array)
+
+        mask_valid_land_pixel = mask_geocoded_burst & mask_land
+
+        percent_valid_land_px = mask_valid_land_pixel.sum() / mask_geocoded_burst.sum() * 100
+        percent_valid_px = mask_valid_inside_burst.sum() / mask_geocoded_burst.sum() * 100
```
Comment (LiangJYu) [inline src/compass/s1_cslc_qa.py:546]:
```suggestion
        n_unmasked_pxls = mask_geocoded_burst.sum()
        percent_valid_land_px = mask_valid_land_pixel.sum() / n_unmasked_pxls * 100
        percent_valid_px = mask_valid_inside_burst.sum() / n_unmasked_pxls * 100
```
cache sum

File: src/compass/utils/h5_helpers.py [hunk 14]
```diff
@@ -21,7 +21,7 @@
 DATA_PATH = '/data'
 QA_PATH = '/quality_assurance'
 METADATA_PATH = '/metadata'
-
+OPERA_OPERATION_CONTACT_EMAIL = 'opera-sds-ops@jpl.nasa.gov'
```
Comment (LiangJYu) [inline src/compass/utils/h5_helpers.py:24]: Since it's not directly HDF5 related, I think this is better in `helper.py`. What do you think?

**Review by LiangJYu — APPROVED** [review]
LGTM. Minor comments and nits

File: src/compass/s1_cslc_qa.py [hunk 15]
```diff
@@ -261,7 +270,7 @@ def shadow_pixel_classification(self, cslc_h5py_root):
                                  pxl_qa_items)
 
 
-    def valid_pixel_percentages(self, cslc_h5py_root):
+    def valid_pixel_percentages(self, cslc_h5py_root, pol):
```
[the remark this answers is not in the document]
Comment (seongsujeong) [reply src/compass/s1_cslc_qa.py:281]: We have another pixel percentage function for layover shadow mask i.e. `shadow_pixel_classification()` so Would like to keep the current function name to avoid confusion.

File: src/compass/s1_cslc_qa.py [hunk 16]
```diff
@@ -270,10 +279,13 @@ def valid_pixel_percentages(self, cslc_h5py_root):
         cslc_h5py_root: h5py.File
             Root of CSLC HDF5
         '''
+
+        percent_land_pixels, percent_valid_pixels = self.compute_valid_land_pixel_percent(cslc_h5py_root,
```
[the remark this answers is not in the document]
Comment (seongsujeong) [reply src/compass/s1_cslc_qa.py:291]: Changed the function name as suggested. Thanks

File: src/compass/s1_cslc_qa.py (hunk 6 above)
Comment (seongsujeong) [reply src/compass/s1_cslc_qa.py:555]: Yes it does. Thanks for the suggestion!

File: src/compass/utils/h5_helpers.py (hunk 14 above)
Comment (seongsujeong) [reply src/compass/utils/h5_helpers.py:24]: Good idea. Thanks for the suggestion :)

File: src/compass/s1_cslc_qa.py [hunk 17]
```diff
@@ -466,3 +478,97 @@ def write_qa_dicts_to_json(self, file_path):
...
+                Path to the coastline shapefile
+
+
+            Returns
+            -------
+            ratio_valid_pixel_land:  float
+                Ratio of valid pixels on land
+        '''
+
+        drv_shp_in = ogr.GetDriverByName('GPKG')
```
[the remark this answers is not in the document]
Comment (seongsujeong) [reply src/compass/s1_cslc_qa.py]: Decided to load the polygons using geopandas so marking this comment as resolved.

File: src/compass/s1_cslc_qa.py [hunk 18]
```diff
@@ -466,3 +478,97 @@ def write_qa_dicts_to_json(self, file_path):
...
+            cslc_h5py_path: h5py.File
+                Root of the CSLC-S1 HDF5 product
+            shapefile_path: str
+                Path to the coastline shapefile
+
+
+            Returns
+            -------
+            ratio_valid_pixel_land:  float
+                Ratio of valid pixels on land
```
[the remark this answers is not in the document]
Comment (seongsujeong) [reply src/compass/s1_cslc_qa.py]: Thanks for the comment. The docsrtings in the functions are revised.

File: src/compass/s1_cslc_qa.py (hunk 1 above)
Comment (seongsujeong) [reply src/compass/s1_cslc_qa.py]: Will keep the code as it is.

File: src/compass/s1_cslc_qa.py [hunk 19]
```diff
@@ -466,3 +488,177 @@ def write_qa_dicts_to_json(self, file_path):
...
+                compute the valid pixel area
+
+            Returns
+            -------
+            percent_valid_land_px: float
+                Percentage of valid pixels on land
+                in the geocoded burst area
+            percent_valid_px: float
+                Percentage of invalid pixels
+                in the geocoded burst area
```
Comment (LiangJYu) [inline src/compass/s1_cslc_qa.py:512]: Docstring off by one four extra spaces.

File: src/compass/s1_cslc_qa.py [hunk 20]
```diff
@@ -466,3 +488,177 @@ def write_qa_dicts_to_json(self, file_path):
...
+            -------
+            percent_shadow: float
+                Percentage of the shadow pixels
+                in the geocoded burst area
+            percent_layover: float
+                Percentage of the layover pixels
+                in the geocoded burst area
+            percent_combined: float
+                Percentage of the shadow and layover pixels
+                in the geocoded burst area
```
Comment (LiangJYu) [inline src/compass/s1_cslc_qa.py:565]: Entire docstring off by extra four spaces

File: src/compass/s1_cslc_qa.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (LiangJYu) [reply src/compass/s1_cslc_qa.py:280]: `percent_land_and_valid_pixels`? Verbose but specific.

File: src/compass/s1_cslc_qa.py [hunk 21]
```diff
@@ -466,3 +488,177 @@ def write_qa_dicts_to_json(self, file_path):
...
+        y0 = list(cslc_h5py_root[f'{DATA_PATH}/y_coordinates'][()])[0] - y_spacing / 2
+
+        cslc_array = np.array(cslc_h5py_root[f'{DATA_PATH}/{pol}'])
+
+        height_cslc, width_cslc = cslc_array.shape
+
+        mask_land = _get_land_mask(epsg_cslc,
+                                   (x0, x_spacing, 0, y0, 0, y_spacing),
+                                   (height_cslc, width_cslc))
+
```
Comment (LiangJYu) [inline src/compass/s1_cslc_qa.py:530]:
```suggestion
```
formatting nit - extra line

**Review by vbrancat — APPROVED** [review]
LGTM. Please, solve the codacy issues. Thanks

File: src/compass/s1_cslc_qa.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (seongsujeong) [reply src/compass/s1_cslc_qa.py:280]: @LiangJYu That looks like a good name. Changing the function name to your suggestion.

**Comment by seongsujeong** [comment]
Thank you all for the quick turnaround and insightful comments. I'll merge this PR as soon as I confirm that it passes the CI test.
